features/202-session-management/design.md

```diff
@@ -0,0 +1,163 @@
+- session ID
+- app ID
+- user ID
+- identity ID
+- expiration time
+- creation time
+- last access time
+- creation IP
+- last IP
+- creation user agent
+- last user agent
+- custom attributes
+
+Session attributes would be available in web-hook event context, and session
+management APIs.
+
+App developer can update custom attributes in session using web-hook event
+mutation.
+
+Gateway is responsible to update last access time, last IP, and last user agent
+of session after processing the request, regardless whether the request is
+successful.
+
+
+## Session Management
+
+APIs would be provided to manage sessions of a user.
+
+Session list API returns a list of valid session of a user. Each list entry
+contains the session ID and session attributes. Access token / refresh token of
+the session would not be included.
+
+Revoke sessions API revokes a specific session, or all sessions other than the
+requesting session. A `session_delete` event would be generated for all sessions
+to be revoked.
+
+
+## Access / Refresh Token
+
+Each session would have an access token, and optionally a refresh token. Both
+type of tokens are opaque: token formats are unspecified and developer should
+not attempt to interpret the content of tokens.
+
+Access token would always have a lifetime; access token would be treated as
+invalid after its expiry.
+
+Refresh token can be used to obtain a new access token. If the old access token
+is still valid, the old access token is invalidated; there would be at most one
+valid access token for a session at any time.
+
+Refresh token would always have a maximum lifetime.
+
+Optionally, a session idle timeout can be specified: refresh token (or access
+token if refresh token is disabled) must be used at least once before the
+timeout, otherwise the session would be expired.
+
+A session is invalidated if its identity is deleted, or its user is disabled.
+A session is expired if its refresh token is expired, or its access token is
+expired (if refresh token is disabled).
+If a session is invalidated, expired, revoked, or logout, its associated access
+token / refresh token would be treated as invalid.
+
+Client-side SDK should handle refreshing of expired access token automatically
+if refresh token is available; SDKs would not expose concept of access token /
+refresh token; however, developers should still handle situation where session
+is no longer valid due to various reasons.
+
+
+## Session Cookie
+
+Optionally, developer can choose to use cookies to transport session tokens, in
+order to facilitate server-side rendering.
+
+If developer choose to enable session cookie:
+- Refresh token would not be issued. (See [Appendix](#session-cookie-and-refresh-token))
+- Access token would be transported in session cookie.
+- Access token would not be returned from APIs.
+- Gateway would read access token from cookie only, ignoring `Authorization`
+  header.
+
+### Security Consideration
+
+- HTTPS is required.
```

Comment:
I guess this is `secure; `. It should be configurable for development purpose and by default it is **enabled**.